iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -320,14 +322,106 @@ static Value buildFlowWorkgroupInfoOp(OpBuilder &b, unsigned dim) {
+      if (allOps.count(definingOp)) leafOps.remove(op);
+    }
+  }
+
+  // The leaves are at the head of the ordered list.
+  SmallVector<Operation *, 16> orderedOps = llvm::to_vector<4>(leafOps);
+  orderedOps.reserve(ops.size());
+  llvm::SmallPtrSet<Operation *, 16> processed;
+  processed.insert(leafOps.begin(), leafOps.end());
+
+  // `readyOps` contains the list of operations that have been just added to the
+  // `orderedOps` list. With these marked ready, they might make further
+  // operations in `ops` ready as well.
+  // The complexity of the algorithm is driven by these
+  // - Each operations is added to `readyOps` list at most once, and is removed
+  //   after being processed
+  // - For every operation in `readyOps` every use of its results (within `ops`)
+  //   is looked at once.
+  // - For every use, the operands of the user are processed.
+  // Assuming operands is O(1), i.e. constant order, the complexity is O(sum of
+  // number of uses of each operation). Given that the size of `ops` is at max
+  // O(10), and not O(100), this is assumed to be reasonable.
+  SmallVector<Operation *, 16> readyOps = orderedOps;
+  readyOps.assign(leafOps.begin(), leafOps.end());
+  while (!readyOps.empty()) {
+    SmallVector<Operation *> nextReadyOps;
+    DEBUG_WITH_TYPE(DEBUG_TYPE, {
+      llvm::dbgs() << "ReadyOps :\n";
+      for (auto readyOp : readyOps) {
+        llvm::dbgs() << "\t";
+        readyOp->print(llvm::dbgs());
+        llvm::dbgs() << "\n";
+      }
+    });
+    for (auto readyOp : readyOps) {
+      // Check all uses of `readyOp` within `ops`. If all of the operations that
+      // define the operands of the user have been added to `orderedOps`, then
+      // the user is ready to be scheduled.
+      for (auto insertAfterOp : insertAfterMap[readyOp]) {
+        if (processed.count(insertAfterOp)) continue;
+        if (llvm::all_of(insertAfterOp->getOperands(), [&](Value operand) {
+              Operation *operandDefiningOp = operand.getDefiningOp();
+              return !operandDefiningOp || processed.count(operandDefiningOp);
+            })) {
+          nextReadyOps.push_back(insertAfterOp);
+          orderedOps.push_back(insertAfterOp);
+          processed.insert(insertAfterOp);
+        }
+      }
+    }
+    std::swap(nextReadyOps, readyOps);
+  }
+
+  DEBUG_WITH_TYPE(DEBUG_TYPE, {
+    llvm::dbgs() << "Ops to be inlined (sorted) : \n";
+    for (auto op : orderedOps) {
+      llvm::dbgs() << "\t";
+      op->print(llvm::dbgs());
+      llvm::dbgs() << "\n";
+    }
+  });
+  assert(orderedOps.size() == ops.size());
+  return std::move(orderedOps);
+}
+
 /// Computes the values that will be eventually be used within the dispatch
 /// workgroup op but defined outside the op after all clonable operations are
 /// cloned into the region. Returns (by reference) the clonable operations too,
 /// in order in which they can be cloned within the region to satisfy use-def
 /// relationships between them.
```

Comment:
I don't think the function returns anything... should we update the comment? It seems that we store the results into `valuesDefinedAbove` (by reverse it + prepend clonedOps).

Reply:
The comment says "returns by reference".